processing/src/main/java/io/druid/segment/BaseObjectColumnValueSelector.java

```diff
@@ -33,6 +33,17 @@
 @PublicApi
 public interface BaseObjectColumnValueSelector<T>
 {
+  /**
+   * Returns the column value at the current position.
+   *
+   * IMPORTANT. The returned object could generally be reused inside the implementation of
+   * BaseObjectColumnValueSelector, i. e. this method could always return the same object for the same selector. Users
+   * of this API, such as {@link io.druid.query.aggregation.Aggregator#aggregate()}, {@link
+   * io.druid.query.aggregation.BufferAggregator#aggregate}, {@link io.druid.query.aggregation.AggregateCombiner#reset},
+   * {@link io.druid.query.aggregation.AggregateCombiner#fold} should be prepared for that and not storing the object
+   * returned from this method in their state, assuming that the object will remain unchanged even when the position of
+   * the selector changes. This may not be the case.
+   */
```

Comment:
Looks like this new contract might this break SelectQuery as there we just call getObject and put that object in an event ? 
 https://github.com/druid-io/druid/blob/535ec437e925effee920e5643277fde2a1b69175/processing/src/main/java/io/druid/query/select/SelectQueryEngine.java#L313

Reply:
Thanks for this finding. I think I will need to add a method like `getObjectNotReusable()` to `BaseObjectColumnValueSelector` and use it in `SelectQueryEngine` in one of the future PRs, where object reuse is actually implemented.

Reply:
1) I am worried about changing the api semantics for existing apis in a backwards incompatible way. In this case if any extensions are using this api, they might break without any compilation error. I think a better way might be to add another api which returns some Holder object. 
2) Also, IMO api that reuse objects internally are a bit tricky and are quite fragile to implement. Another way could be to use Druid Sequence or a similar api which do not expose the underlying object and provide a way to accumulate or apply computations to the object.

Reply:
1. This is fair, while developing the actual PR that makes getObject() to return reusable objects I came to the same conclusion, and decided to also rename the "main" method to getReusableObject(), and the second is getNotReusableObject().

Using any "holder" interlayer won't allow to bypass the fact that we inherently going to reuse the same, reusable objects. So I don't see the point of adding such.

Reply:
2. Implementors of custom `ObjectStrategy` could always retreat to use not reusable objects, it just won't be that efficient when a column with their `ObjectStrategy` is used. Users of this API could also retreat to always use `getNotReusableObject()` if they are unsure.

I don't say that introduction of object reusability doesn't add complexity, it does, but there are no interfaces that allow to hide this complexity, `Sequence`, `ColumnValueSelector` or any else, because it's still about making immutable objects mutable. Not to say that rewriting everything from `ColumnValueSelector` to `Sequence` would be a MASSIVE refactoring and it's absolutely not guaranteed that it will make anything clearer. So it doesn't seem realistic to me.

Reply:
In this PR, I removed this doc update of `getObject()`, because it's premature, this PR is about `IndexedInts` and `DimensionSelector`s. Let's focus on it.